### DISCUSSION TITLE
[Nextcloud] Install both Rainloop and Snappymail

### DISCUSSION THREAD

**Discussion by cm-schl** [the-djmaze/snappymail#583] [Q&A] [answered]
Hi,

after testing SnappyMail in a Nextcloud test environment I would like to activate the SnappyMail App in a Nextcloud Installation that also has already installed Rainloop. I then would activate SnappyMail for some users (for testing) and at the same time leave Rainloop active for others.
I saw, that SnappyMail imports settings from existing Rainloop installs so I think this is possible. But I want to ask if this can be done without generating unexpected problems to the users that are actively using Rainloop?

**Comment by cm-schl** [comment]
I've set up a complete copy of my environment to be sure what happens when Rainloop and SnappyMail is installed at the same time.
As far as I see the first time the admin enters into the "Additional settings" of Nextcloud the importRainLoop function is fired https://github.com/the-djmaze/snappymail/blob/3903a37b647aea788ebc64b54b3225843fc9886e/integrations/nextcloud/snappymail/lib/Util/SnappyMailHelper.php#L137 and copies everything from the Rainloop folder on the filesystem to SnappyMail. 
This seems to happen only once and without any info to the admin - the password of the SnappyMail Admin Interface in this way is the same as the password of the Rainloop Admin Interface. That was irritating for a moment 🙂 .

So in theory it looks like it would possible to use both without any problems and migrate the users after everything has been set up.

@the-djmaze could you eventually confirm / correct my assertion? 🙂

off topic: I don't get what the || operator is doing in this line: https://github.com/the-djmaze/snappymail/blob/3903a37b647aea788ebc64b54b3225843fc9886e/integrations/nextcloud/snappymail/lib/Util/SnappyMailHelper.php#L146
I (think I) understand what || does, but as there is no variable where the result can be written to this line isn't clear to me - could somebody help me please? 🙂

**Reply by the-djmaze** [reply]
> and copies everything from the Rainloop folder on the filesystem to SnappyMail.
This seems to happen only once and without any info to the admin - the password of the SnappyMail Admin Interface in this way is the same as the password of the Rainloop Admin Interface. That was irritating for a moment slightly_smiling_face .

Yes, this is changed for next release as people got confused.
Then the page will have only have an import rainloop button when `admin_password.txt` exists.
And it will not overwrite existing SnappyMail files.

> could you eventually confirm / correct my assertion? slightly_smiling_face

Yes mostly it works, but some people also reported that it didn't for them.
Why is always a riddle.

> I don't get what the || operator is doing

is_dir OR mkdir
directory exists ELSE make directory
if directory does not exist then make directory

examples:
```php
if (!\is_dir($dir_snappy)) {
	\mkdir($dir_snappy, 0755, true);
}

if (\is_dir($dir_snappy) || \mkdir($dir_snappy, 0755, true)) {
	echo 'Directory exists or created';
}

if (!\is_dir($dir_snappy) && !\mkdir($dir_snappy, 0755, true)) {
	echo 'Directory does not exist and create failed';
}
```

**Reply by cm-schl** [reply]
Thanks a lot @the-djmaze for all your work and also for the explanation 🙂.

**Comment by cm-schl** [comment, marked as answer]
Just because it also happened to me and maybe someone comes into the same situation: in the meantime SnappyMail and Rainloop can't be installed at the same time in a Nextcloud - you can do this to migrate data from Rainloop to SnappyMail, but a productive usage of both (maybe for some days to migrate everything etc.) isn't possible. Please see also: https://github.com/the-djmaze/snappymail/issues/660

**Reply by Niveshkrishna** [reply]
@cm-schl Could you please tell me what issues you have faced when having both apps enabled in production, apart from missing version directory error? Also have you tried changing namespace in rainloop app so that it does not interfere with snappymail?

**Reply by cm-schl** [reply]
at the moment the only problem I saw was the "missing version" error on logout. Other problems I did not see - maybe there are some others.
Didn't tried to change the namespace - in my situation I enabled and disabled SnappyMail inside the Nextcloud Apps until everything was configured... When an app gets disabled the configuration remains, so you do not loose your settings.
